plugins/hls-completions-plugin/src/Ide/Plugin/LocalCompletions.hs

```diff
@@ -0,0 +1,705 @@
+    CC a <> CC a' =
+        CC (a<>a')
+
+
+data CompItem = CI
+  { compKind :: CompletionItemKind,
+    -- | Snippet for the completion
+    insertText :: T.Text,
+    -- | From where this item is imported from.
+    importedFrom :: Either SrcSpan T.Text,
+    -- | Available type information.
+    typeText :: Maybe T.Text,
+    -- | Label to display to the user.
+    label :: T.Text,
+    -- | Did the completion happen
+    -- in the context of an infix notation.
+    isInfix :: Maybe Backtick,
+    -- | Available documentation.
+    docs :: SpanDoc,
+    isTypeCompl :: Bool,
+    additionalTextEdits :: Maybe [TextEdit]
+  }
+  deriving (Eq, Show)
+
+-- ---------------------------------------------------------------------
+-- Generating Local Completions via Rules
+-- ---------------------------------------------------------------------
+
+produceLocalCompletions :: Rules ()
+produceLocalCompletions = do
+    define $ \LocalCompletions file -> do
+        pm <- useWithStale GetParsedModule file
+        case pm of
+            Just (pm, _) -> do
+                let cdata = localCompletionsForParsedModule pm
+                return ([], Just cdata)
+            _ -> return ([], Nothing)
+
+-- | Produce completions info for a file
+type instance RuleResult LocalCompletions = CachedCompletions
+
+data LocalCompletions = LocalCompletions
+     deriving (Eq, Show, Typeable, GG.Generic)
+instance Hashable LocalCompletions
+instance NFData   LocalCompletions
+instance Binary   LocalCompletions
+
+-- | Generate code actions.
+getCompletionsLSP
+    :: LSP.LspFuncs cofd
+    -> IdeState
+    -> CompletionParams
+    -> IO (Either ResponseError CompletionResponseResult)
+getCompletionsLSP lsp ide
+  CompletionParams{_textDocument=TextDocumentIdentifier uri
+                  ,_position=position
+                  ,_context=completionContext} = do
+    contents <- LSP.getVirtualFileFunc lsp $ toNormalizedUri uri
+    fmap Right $ case (contents, uriToFilePath' uri) of
+      (Just cnts, Just path) -> do
+        let npath = toNormalizedFilePath' path
+        (ideOpts, compls) <- runIdeAction "Completion" (shakeExtras ide) $ do
+            opts <- liftIO $ getIdeOptionsIO $ shakeExtras ide
+            compls <- useWithStaleFast LocalCompletions npath
+            pm <- useWithStaleFast GetParsedModule npath
+            binds <- fromMaybe (mempty, zeroMapping) <$> useWithStaleFast GetBindings npath
+            pure (opts, fmap (,pm,binds) compls )
+        case compls of
+          Just ((cci', _), parsedMod, bindMap) -> do
+            pfix <- VFS.getCompletionPrefix position cnts
+            case (pfix, completionContext) of
+              (Just (VFS.PosPrefixInfo _ "" _ _), Just CompletionContext { _triggerCharacter = Just "."})
+                -> return (Completions $ List [])
+              (Just pfix', _) -> do
+                  let extras = shakeExtras ide
+                  let clientCaps = S.clientCapabilities extras
```

Comment:
@pepeiborra or others. I could use a helping hand here.  I am not able to figure out why I get this error here. I see that 'Development.IDE.Core.Shake’ is an exposed module and `ShakeExtras(..)` is imported. But, still, I see this error:

```
hls-completions-plugin > /home/devanla/fsf/hls-dev/plugins/hls-completions-plugin/src/Ide/Plugin/LocalCompletions.hs:198:36: error:
hls-completions-plugin >     Not in scope: ‘S.clientCapabilities’
hls-completions-plugin >     Perhaps you meant ‘LSP.clientCapabilities’ (imported from Language.Haskell.LSP.Core)
hls-completions-plugin >     Module ‘Development.IDE.Core.Shake’ does not export ‘clientCapabilities’.
hls-completions-plugin >     |
hls-completions-plugin > 198 |                   let clientCaps = S.clientCapabilities extras
hls-completions-plugin >     |                                    ^^^^^^^^^^^^^^^^^^^^

``
Perhaps, some other settings is overriding this in hls. The same expression seems to work fine from inside `ghcide`.

Reply:
Did you figure this out?

Reply:
Nope. I have not spent time on this since then. Plan to look at in again soon. But, if you have any insight please do share.

Reply:
I have been able to resolve this issue. I just need a later version of ghcide where this field was added to ShakeExtras.